cvmfs/catalog_balancer_impl.h

```diff
@@ -0,0 +1,189 @@
+#include <string>
+#include <vector>
+
+#include "catalog_mgr.h"
+#include "directory_entry.h"
+#include "hash.h"
+#include "logging.h"
+
+
+using namespace std;  // NOLINT
+
+namespace catalog {
+
+template <class CatalogMgrT>
+DirectoryEntryBase
+CatalogBalancer<CatalogMgrT>::MakeEmptyDirectoryEntryBase(
+    string name, uid_t uid, gid_t gid) {
+  shash::Algorithms algorithm = catalog_mgr_->spooler_->GetHashAlgorithm();
+  shash::Any file_hash(algorithm);
+  shash::HashString("", &file_hash);
+
+  DirectoryEntryBase deb;
+  deb.name_ = NameString(name);
+  deb.mode_ = S_IFREG | S_IRWXU;
+  deb.checksum_ = file_hash;
+  deb.mtime_ = time(NULL);
+  deb.uid_ = uid;
+  deb.gid_ = gid;
+  return deb;
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::AddCatalogMarker(string path) {
+  XattrList xattr;
+  DirectoryEntry parent;
+  bool retval;
+  retval = catalog_mgr_->LookupPath(PathString(path), kLookupSole, &parent);
+  assert(retval);
+  DirectoryEntryBase cvmfscatalog =
+      MakeEmptyDirectoryEntryBase(".cvmfscatalog", parent.uid(),
+                                           parent.gid());
+  DirectoryEntryBase cvmfsautocatalog =
+      MakeEmptyDirectoryEntryBase(".cvmfsautocatalog", parent.uid(),
+                                           parent.gid());
+  string relative_path = path.substr(1);
+  catalog_mgr_->AddFile(cvmfscatalog, xattr, relative_path);
+  catalog_mgr_->AddFile(cvmfsautocatalog, xattr, relative_path);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::Balance(catalog_t *catalog) {
+  if (catalog == NULL) {
+    // obtain a copy of the catalogs
+    vector<catalog_t*> catalogs = catalog_mgr_->GetCatalogs();
+    // we need to reverse the catalog list in order to analyze the
+    // last added ones first. This is necessary in the weird case the child
+    // catalog's underflow provokes an overflow in the father
+    reverse(catalogs.begin(), catalogs.end());
+    for (unsigned i = 0; i < catalogs.size(); ++i)
+      Balance(catalogs[i]);
+    return;
+  }
+  string catalog_path = catalog->path().ToString();
+  virtual_node_t root_node(catalog_path, catalog_mgr_);
+  root_node.ExtractChildren(catalog_mgr_);
+  // we have just recursively loaded the entire virtual tree!
+  PartitionOptimally(&root_node);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::PartitionOptimally(
+    virtual_node_t *virtual_node) {
+  // postorder track of the fs-tree
+  for (unsigned i = 0; i < virtual_node->children.size(); ++i) {
+    virtual_node_t *virtual_child =
+        &virtual_node->children[i];
+    if (virtual_child->IsDirectory() && !virtual_child->IsCatalog())
+      PartitionOptimally(virtual_child);
+  }
+  virtual_node->FixWeight();
+  while (virtual_node->weight > catalog_mgr_->balance_weight_) {
+    virtual_node_t *heaviest_node =
+        MaxChild(virtual_node);
+    // we directly add a catalog in this node
+    // TODO(molina) apply heuristics here
```

Comment:
@reneme and I had some ideas to apply here that would require a bit of whiteboard to be explained, but basically is the following:
1) Make the root catalog smaller than the rest, if necessary. We would have to establish a "heuristic" to set the final maximum weight of the root catalog.
2) If a folder is a nested catalog, it's likely that a set of siblings are as well. Say, if /releases/root/5.4.3 is a catalog, then /releases/root/5.x.x are likely to be a catalog. More concretely, we thought that:
```
weight(FOLDER) >= weight(MAX_FOLDER) * c   then FOLDER should be a nested catalog as well,      with 0 < c < 1 for all FOLDER sibling of MAX_FOLDER
```